pip/pip-310.md

```diff
@@ -0,0 +1,121 @@
+
+# Detailed Design
+
+## Design & Implementation Details
+
+* In `AbstractTopic.java`, instead of if-else based constructor call to either `PublishRateLimiterImpl` or `
+  PrecisePublishLimiter`, we read the `topicPublishRateLimiterProvider` config and load that class instead.
+  * Similarly, in `ResourceGroup.java`, we load the class defined in `resourceGroupPublishRateLimiterProvider` config.
+* Replace constructor call for `ResourceGroupPublishLimiter`, `PrecisePublishLimiter` and `PublishRateLimiterImpl` to `.initialize` method call.
+* Add fallback & default logic to load `PublishRateLimiterImpl` and `ResourceGroupPublishLimiter` in case of ClassNotFoundException or exception in `initialize` call in `AbstractTopic.java` and `
+  ResourceGroup.java`
+* For cases where rate limiting is disabled, we continue to use `PublishRateLimiterDisable` like we do currently.
+* For backward compatibility (for 2 minor versions), in `AbstractTopic.java` we add following logic:
+  * In case users have the flag `preciseTopicPublishRateLimiterEnable` as true *and* the `
+    topicPublishRateLimiterProvider` is empty, we still load `PublishRateLimiterImpl`.
+  * In case `topicPublishRateLimiterProvider` is a valid class, we ignore the value of `
+    preciseTopicPublishRateLimiterEnable`
+  * We mark the config `preciseTopicPublishRateLimiterEnable` deprecated in the config file and ask users to use the provider config with the classpath.
+
+### Configuration
+
+* `topicPublishRateLimiterProvider: <classpath>`
+* `resourceGroupPublishRateLimiterProvider: <classpath>`
+
+# Backward & Forward Compatibility
+
+## Upgrade
+
+In detailed design, backward compatibility logic is explained. Upgrades would work without need of config change until
+we finally remove the logic to understand `preciseTopicPublishRateLimiterEnable`.
+
+As a best practice:
+
+* Users requiring precise rate limiter remove the config `preciseTopicPublishRateLimiterEnable` and
+  add `topicPublishRateLimiterProvider: org.apache.pulsar.broker.service.PrecisePublishLimiter`
+* Users requiring custom rate limiter ensure that their class is present in pulsar classpath and they set
+  the `topicPublishRateLimiterProvider` or `
+  resourceGroupPublishRateLimiterProvider` accordingly
+
+## Revert
+
+In case user was not using the precise rate limiter, simply deleting the two new configs should revert back to previous version.
+
+In case user was using precise rate limiter, i.e., they had `preciseTopicPublishRateLimiterEnable: true` and:
+* They removed that and switched to `topicPublishRateLimiterProvider: org.apache.pulsar.broker.service.PrecisePublishLimiter`
+  * To revert, they simply need to remove the config `topicPublishRateLimiterProvider` and put in `preciseTopicPublishRateLimiterEnable: true`
+* They did not change any of their configs, i.e. they did not specifically add and value in `topicPublishRateLimiterProvider`, then no change needed.
+
+# Links
```

Comment:
You're missing the Public API section

Reply:
I must say I was a bit confused by the purpose of "Public API" section. Who is the "Public" here? I thought that only changes going in the pulsar client handshake related interfaces are Public. I tried to refer to older PIPs but there was no solid conclusion.

In this PIP, will the interface `PublishRateLimiter.java` be the Public API?

Reply:
Take OpenTelemetry java SDK for example - they clearly mark in the package which interface/class is public. I believe in Pulsar we don't have that, but the basic idea remains: You need to document in the PIP what changes you are making to code you intend the public to use, whether introducing a new interface or changing an existing one.  

PublishRateLimiter is public API since Pulsar users will create instances of that interfac and you can't break in the future? Hence it's public API.

Reply:
Noted. thanks for the explanation. Will add in here.